### ISSUE TITLE
Specifying -f s16le in postprocessor-args seems to stop M4AFixup from working

### ISSUE THREAD

**Issue by Metritutus** [yt-dlp/yt-dlp#2766]
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

I am trying to download audio only and convert that audio to PCM.

For some reason, when a format argument (in my case, `-f s16le`) is specified in `--postprocessor-args`, m4a fixup (where it apparently adjusts the file to not be a `DASH` one or something like that) doesn't seem to actually happen, although it says it does and it doesn't give an error.

I've omitted the `--extract-audio --audio-format wav` stuff as that falls over at the end as a result, but I don't think it's connected to this as I can see the m4a file is clearly wrong after the first command has run.

Command 1 (with `-f s16le`)
```
yt-dlp -o "%(id)s.%(ext)s" -f bestaudio/best --postprocessor-args "ffmpeg:-ac 2 -f s16le -ar 48000" --compat-options youtube-dl https://www.youtube.com/watch?v=54rwawswQpM/
```
Command 2 (without `-f s16le`)
```
yt-dlp -o "%(id)s.%(ext)s" -f bestaudio/best --postprocessor-args "ffmpeg:-ac 2 -ar 48000" --compat-options youtube-dl https://www.youtube.com/watch?v=54rwawswQpM/
```

### Verbose log

```shell
Command 1 output:

[debug] Command-line config: ['-o', '%(id)s.%(ext)s', '-f', 'bestaudio/best', '--postprocessor-args', 'ffmpeg:-ac 2 -f s16le -ar 48000', '--compat-options', 'youtube-dl', 'https://www.youtube.com/watch?v=54rwawswQpM/', '-vU']
[debug] Encodings: locale cp1252, fs utf-8, out cp1252 (No ANSI), err cp1252 (No ANSI), pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Compatibility options: no-youtube-unavailable-videos, format-spec, seperate-video-versions, no-attach-info-json, playlist-index, no-youtube-channel-redirect, list-formats, embed-thumbnail-atomicparsley, no-clean-infojson, embed-metadata, no-direct-merge, *filename, no-playlist-metafiles, no-keep-subs, abort-on-error, format-sort, no-live-chat
[debug] Python version 3.8.10 (CPython 64bit) - Windows-8.1-6.3.9600-SP0
[debug] exe versions: ffmpeg 5.0-full_build-www.gyan.dev (setts), ffprobe 5.0-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=54rwawswQpM/
[youtube] 54rwawswQpM: Downloading webpage
[youtube] 54rwawswQpM: Downloading android player API JSON
[debug] Sort order given by user: hasaud, lang, quality, tbr, filesize, vbr, height, width, proto, vext, abr, aext, fps, fs_approx, source, id
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, hasaud, lang, quality, tbr, filesize, vbr, height, width, proto, vext, abr, aext, fps, fs_approx, source, id, res, hdr:12(7), vcodec:vp9.2(10), acodec, asr
[info] 54rwawswQpM: Downloading 1 format(s): 140
[debug] Invoking downloader on "https://rr5---sn-aigl6nl7.googlevideo.com/videoplayback?<REDACTED>"
[download] Destination: 54rwawswQpM.m4a

[download]   0.0% of 5.29MiB at Unknown speed ETA Unknown
[download]   0.1% of 5.29MiB at Unknown speed ETA Unknown
[download]   0.1% of 5.29MiB at Unknown speed ETA Unknown
[download]   0.3% of 5.29MiB at  7.50MiB/s ETA 00:00     
[download]   0.6% of 5.29MiB at  3.87MiB/s ETA 00:01
[download]   1.2% of 5.29MiB at  4.85MiB/s ETA 00:01
[download]   2.3% of 5.29MiB at  6.05MiB/s ETA 00:00
[download]   4.7% of 5.29MiB at  7.97MiB/s ETA 00:00
[download]   9.4% of 5.29MiB at 11.61MiB/s ETA 00:00
[download]  18.9% of 5.29MiB at 15.98MiB/s ETA 00:00
[download]  37.8% of 5.29MiB at 18.95MiB/s ETA 00:00
[download]  75.6% of 5.29MiB at 18.36MiB/s ETA 00:00
[download] 100% of 5.29MiB at 18.43MiB/s ETA 00:00  
[download] 100% of 5.29MiB in 00:00               
[FixupM4a] Correcting container of "54rwawswQpM.m4a"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:54rwawswQpM.m4a" -map 0 -dn -ignore_unknown -c copy -f mp4 -movflags "+faststart" -ac 2 -f s16le -ar 48000 "file:54rwawswQpM.temp.m4a"

Command 2 output:

[debug] Command-line config: ['-o', '%(id)s.%(ext)s', '-f', 'bestaudio/best', '--postprocessor-args', 'ffmpeg:-ac 2 -ar 48000', '--compat-options', 'youtube-dl', 'https://www.youtube.com/watch?v=54rwawswQpM/', '-vU']
[debug] Encodings: locale cp1252, fs utf-8, out cp1252 (No ANSI), err cp1252 (No ANSI), pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Compatibility options: no-attach-info-json, embed-metadata, embed-thumbnail-atomicparsley, no-youtube-unavailable-videos, no-clean-infojson, no-direct-merge, list-formats, no-youtube-channel-redirect, no-keep-subs, *filename, no-playlist-metafiles, seperate-video-versions, playlist-index, format-spec, no-live-chat, format-sort, abort-on-error
[debug] Python version 3.8.10 (CPython 64bit) - Windows-8.1-6.3.9600-SP0
[debug] exe versions: ffmpeg 5.0-full_build-www.gyan.dev (setts), ffprobe 5.0-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=54rwawswQpM/
[youtube] 54rwawswQpM: Downloading webpage
[youtube] 54rwawswQpM: Downloading android player API JSON
[debug] Sort order given by user: hasaud, lang, quality, tbr, filesize, vbr, height, width, proto, vext, abr, aext, fps, fs_approx, source, id
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, hasaud, lang, quality, tbr, filesize, vbr, height, width, proto, vext, abr, aext, fps, fs_approx, source, id, res, hdr:12(7), vcodec:vp9.2(10), acodec, asr
[info] 54rwawswQpM: Downloading 1 format(s): 140
[debug] Invoking downloader on "https://rr5---sn-aigzrne7.googlevideo.com/videoplayback?<REDACTED>"
[download] Destination: 54rwawswQpM.m4a

[download]   0.0% of 5.29MiB at 341.28KiB/s ETA 00:15
[download]   0.1% of 5.29MiB at 1023.83KiB/s ETA 00:05
[download]   0.1% of 5.29MiB at  2.33MiB/s ETA 00:02  
[download]   0.3% of 5.29MiB at  5.00MiB/s ETA 00:01
[download]   0.6% of 5.29MiB at  3.87MiB/s ETA 00:01
[download]   1.2% of 5.29MiB at  4.85MiB/s ETA 00:01
[download]   2.3% of 5.29MiB at  6.05MiB/s ETA 00:00
[download]   4.7% of 5.29MiB at  7.73MiB/s ETA 00:00
[download]   9.4% of 5.29MiB at 11.61MiB/s ETA 00:00
[download]  18.9% of 5.29MiB at 15.50MiB/s ETA 00:00
[download]  37.8% of 5.29MiB at 17.20MiB/s ETA 00:00
[download]  75.6% of 5.29MiB at 17.80MiB/s ETA 00:00
[download] 100% of 5.29MiB at 17.94MiB/s ETA 00:00  
[download] 100% of 5.29MiB in 00:00               
[FixupM4a] Correcting container of "54rwawswQpM.m4a"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:54rwawswQpM.m4a" -map 0 -dn -ignore_unknown -c copy -f mp4 -movflags "+faststart" -ac 2 -ar 48000 "file:54rwawswQpM.temp.m4a"
```

**Comment by pukkandan** [comment]
> [debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:54rwawswQpM.m4a" -map 0 -dn -ignore_unknown -c copy -f mp4 -movflags "+faststart" -ac 2 -f s16le -ar 48000 "file:54rwawswQpM.temp.m4a"


It is running. 

Even if it wasn't, that is not a bug. The fixup processes may be skipped if we determine it is not required. It is a valid issue only if the resulting file has any issues

**Comment by pukkandan** [comment]
It is a user error. The fixup step is running, and the arguments you asked to be added are being added. But you are asking to put `pcm_s16le` into an m4a file (which is not valid afaik). So the file ends up damaged

**Comment by Zirro** [comment]
There are two postprocessors involved here - FixupM4a and ExtractAudio. I think you'll want to use `--postprocessor-args 'ExtractAudio:-ac 2 -ar 48000'` to only pass the arguments to the latter instead.

I also removed the `-f` flag since s16le appears to be the default for WAV files, and doesn't seem to be an accepted value for this flag either way. Perhaps what you were looking for was `-c:a pcm_s16le`?

Here's the full command with modifications:

`yt-dlp -o '%(id)s.%(ext)s' -f 'bestaudio[ext=m4a]' --postprocessor-args 'ExtractAudio:-ac 2 -ar 48000' --compat-options 'youtube-dl' 'https://www.youtube.com/watch?v=54rwawswQpM' -vU --extract-audio --audio-format wav`

**Comment by Metritutus** [comment]
> It is a user error. The fixup step is running, and the arguments you asked to be added are being added. But you are asking to put `pcm_s16le` into an m4a file (which is not valid afaik). So the file ends up damaged

 As stated in the original post, my intent was to ultimately apply such a format to the final PCM WAV file, _not_ the M4A file, however it would apparently apply the ffmpeg post-processor stuff prior to the extract-audio/audio-format stages.

> There are two postprocessors involved here - FixupM4a and ExtractAudio. I think you'll want to use `--postprocessor-args 'ExtractAudio:-ac 2 -ar 48000'` to only pass the arguments to the latter instead.
> 
> I also removed the `-f` flag since s16le appears to be the default for WAV files, and doesn't seem to be an accepted value for this flag either way. Perhaps what you were looking for was `-c:a pcm_s16le`?
> 
> Here's the full command with these modifications:
> 
> `yt-dlp -o '%(id)s.%(ext)s' -f 'bestaudio/best' --postprocessor-args 'ExtractAudio:-c:a pcm_s16le -ac 2 -ar 48000' --compat-options 'youtube-dl' 'https://www.youtube.com/watch?v=54rwawswQpM' -vU --extract-audio --audio-format wav`
> 
> EDIT: `-f` does seem to take `s16le` under some circumstances, after all. This page appears to have examples for what you're trying to achieve (using both `-f` and `-c:a`): https://trac.ffmpeg.org/wiki/audio%20types

Now _this_ was helpful!  This is _exactly_ what I was looking to accomplish, thanks!  I had no idea that ffmpeg arguments could be applied to the ExtractAudio postprocessor.

I will confess I had great difficulty trying to figure out how to accomplish what I wanted here on my own.  _User error,_ as @pukkandan points out, although I feel it worth mentioning that I did try to seek out documentation for how to the postprocessor args actually worked, as well as Googling for examples.  Information seemed somewhat lacking, and Googling "--postprocessor-args" would generally only find a meagre 1500ish results, which were not too illuminating.

Whilst my problem is solved, I feel that it would be helpful if the documentation could perhaps be updated to elaborate on what parameters apply to which postprocessors.  ie if a post processor is an ffmpeg one, then ffmpeg arguments apply.  I'm assuming here that they aren't _all_ ffmpeg ones, as the documentation currently there mentions support for using other executables as postprocessors, and I presume some of the others may well be handled by those applications.

It would also be useful to know what order postprocessors are applied in (ie the ffmpeg one clearly gets applied prior to the ExtractAudio one).

**Comment by pukkandan** [comment]
> however it would apparently apply the ffmpeg post-processor stuff prior to the extract-audio/audio-format stages.

Yes, by giving `ffmpeg:`, you asked it to be run for ALL ffmpeg PP runs. You never mentioned to me that you were not intending to pass the args to the fixupm4a step. To the contrary, your OP that "Specifying -f s16le seems to stop M4AFixup", along with removing the -x from the logs as "not connected" gave me the impression that your are in fact trying to pass those args to the fixup step. That is why I didn't think to point you to using `ExtractAudio+ffmpeg:` prefix

> I feel that it would be helpful if the documentation could perhaps be updated to elaborate on what parameters apply to which postprocessors. ie if a post processor is an ffmpeg one, then ffmpeg arguments apply. I'm assuming here that they aren't all ffmpeg ones, as the documentation currently there mentions support for using other executables as postprocessors, and I presume some of the others may well be handled by those applications.

```
    --postprocessor-args NAME:ARGS   Give these arguments to the postprocessors.
                                     Specify the postprocessor/executable name
                                     and the arguments separated by a colon ":"
                                     to give the argument to the specified
                                     postprocessor/executable. Supported PP are:
                                     Merger, ModifyChapters, SplitChapters,
                                     ExtractAudio, VideoRemuxer, VideoConvertor,
                                     Metadata, EmbedSubtitle, EmbedThumbnail,
                                     SubtitlesConvertor, ThumbnailsConvertor,
                                     FixupStretched, FixupM4a, FixupM3u8,
                                     FixupTimestamp and FixupDuration. The
                                     supported executables are: AtomicParsley,
                                     FFmpeg and FFprobe. You can also specify
                                     "PP+EXE:ARGS" to give the arguments to the
                                     specified executable only when being used
                                     by the specified postprocessor.
                                     Additionally, for ffmpeg/ffprobe, "_i"/"_o"
                                     can be appended to the prefix optionally
                                     followed by a number to pass the argument
                                     before the specified input/output file. Eg:
                                     --ppa "Merger+ffmpeg_i1:-v quiet". You can
                                     use this option multiple times to give
                                     different arguments to different
                                     postprocessors. (Alias: --ppa)
```

This is not clear enough? How would you suggest I improve it?

**Comment by Metritutus** [comment]
> > however it would apparently apply the ffmpeg post-processor stuff prior to the extract-audio/audio-format stages.
> 
> Yes, by giving `ffmpeg:`, you asked it to be run for ALL ffmpeg PP runs.

Which was something I did not realise.  It doesn't seem to be explicitly stated in the documentation.  The only line close to this is the one about the ability `to give the arguments to the specified executable only when being used by the specified postprocessor.`

My assumption at the time was that it was running ffmpeg at the end with the arguments I specified, not that it was running it for _all_ ffmpeg postprocessors (whichever those were).

> > As stated in the original post, my intent was to ultimately apply such a format to the final PCM WAV file, not the M4A file
> 
> You mentioned no such thing!

To quote my original post:

> I am trying to download audio only and convert that audio to PCM.

And further down:
> I've omitted the `--extract-audio --audio-format wav` stuff as that falls over at the end as a result, but I don't think it's connected to this as I can see the m4a file is clearly wrong after the first command has run.

I confess I was not super explicit about this, but only because I didn't think it related to the problem I was apparently experiencing.  In an effort to make the Issue useful and specific, I attempted to remove anything apparently superfluous (identified through testing) to what the problem appeared to be.

> You never even mentioned that you were not intending to pass the args to the fixupm4a step.

I did not mention it because I did not know what arguments I _could_ pass to the postprocessors (such as with extract-audio, which @Zirro kindly made quite clear), nor did I know that, by specifying `ffmpeg`, it was implicitly applying to the FixUpM4A step (in fact I had not even realised it was one of the postprocessors as I missed it when reading the documentation).  It was not a lack of intent so much as it was a lack of knowledge.

> To the contrary, your OP that "Specifying -f s16le seems to stop M4AFixup", along with removing the -x from the logs as "not connected" gave me the impression that your are in fact trying to pass those args to the fixup step. That is why I didn't think to point you to using `ExtractAudio+ffmpeg:` prefix

I inferred they were not connected because the extract-audio step would fall over because the M4A file was in an incorrect format.  From this I presumed at the time that the M4A fixup step had failed to apply prior to the WAV conversion, which I had assumed would then be followed by the ffmpeg postprocessing args I'd specified.

In addition, where I specified 'FixupM4a' before, I merely copied the name from the debug log.  As mentioned, I had missed that it was listed among the other post processors, and just assumed it was an internal operation it happened to be logging.

> > I feel that it would be helpful if the documentation could perhaps be updated to elaborate on what parameters apply to which postprocessors. ie if a post processor is an ffmpeg one, then ffmpeg arguments apply. I'm assuming here that they aren't all ffmpeg ones, as the documentation currently there mentions support for using other executables as postprocessors, and I presume some of the others may well be handled by those applications.
> 
> ```
>     --postprocessor-args NAME:ARGS   Give these arguments to the postprocessors.
>                                      Specify the postprocessor/executable name
>                                      and the arguments separated by a colon ":"
>                                      to give the argument to the specified
>                                      postprocessor/executable. Supported PP are:
>                                      Merger, ModifyChapters, SplitChapters,
>                                      ExtractAudio, VideoRemuxer, VideoConvertor,
>                                      Metadata, EmbedSubtitle, EmbedThumbnail,
>                                      SubtitlesConvertor, ThumbnailsConvertor,
>                                      FixupStretched, FixupM4a, FixupM3u8,
>                                      FixupTimestamp and FixupDuration. The
>                                      supported executables are: AtomicParsley,
>                                      FFmpeg and FFprobe. You can also specify
>                                      "PP+EXE:ARGS" to give the arguments to the
>                                      specified executable only when being used
>                                      by the specified postprocessor.
>                                      Additionally, for ffmpeg/ffprobe, "_i"/"_o"
>                                      can be appended to the prefix optionally
>                                      followed by a number to pass the argument
>                                      before the specified input/output file. Eg:
>                                      --ppa "Merger+ffmpeg_i1:-v quiet". You can
>                                      use this option multiple times to give
>                                      different arguments to different
>                                      postprocessors. (Alias: --ppa)
> ```
> 
> This is not clear enough? How would you suggest I improve it?

First, separate paragraphs would be extremely useful, as right now it's just a wall of text, which makes it easy to miss key nuggets of information held within.

Second, it is not mentioned which postprocessors are apparently applied by which executables, nor is it mentioned that specifying the postprocessing options for an executable applies to all the postprocessors which run that executable.

I did not make the connection between `only when used by the specified postprocessor` and that the fact that the post processors are apparently all handled exclusively by the executables listed for postprocessing.  I had assumed they were being listed as supported as separate additions to the postprocessors that could be configured, not that they were also handling the other postprocessing operations as well, and that they could be configured en-mass by specifying just the executable, as I had done.  It's also unclear if a postprocessor operation can involve more than one executable.

I had also previously assumed that the specified postprocessors were all unique operations specific to yt-dlp (ie not separate executables), and with no supported arguments for each specified in the documentation, I had no idea they used what arguments could be applied to them.  Considering the lack of examples within the documentation or Google results, I didn't exactly have much to go on.  Once this information is known to someone, it may seem obvious to them, however to the initiated, such as myself, it seemed somewhat opaque.

Third, the documentation does not mention the order the post processors are run in.  Does the order the parameters are passed in matter?  What about the order of the ones that run automatically, such as M4AFixup?  Or is the order always fixed regardless of whether the user manually specifies yt-dlp postprocessor parameters (such as `--extract-audio`) or not, or what order those parameters are passed in?  I'd _assume_ that the last one is the case (that the order is fixed), but as should be apparent at this point, assumptions can often be incorrect, as mine have been here.



With regards to my first point regarding paragraphs, as an example of how much more readable paragraphs can make this, I have demonstrated below:

> Give these arguments to the postprocessors. Specify the postprocessor/executable name and the arguments separated by a colon ":" to give the argument to the specified postprocessor/executable.
> 
> Supported PP are:
>   * Merger
>   * ModifyChapters
>   * SplitChapters
>   * ExtractAudio
>   * VideoRemuxer
>   * VideoConvertor
>   * Metadata
>   * EmbedSubtitle
>   * EmbedThumbnail
>   * SubtitlesConvertor
>   * ThumbnailsConvertor
>   * FixupStretched
>   * FixupM4a
>   * FixupM3u8
>   * FixupTimestamp
>   * FixupDuration
>   
> The supported executables are:
>   * AtomicParsley
>   * FFmpeg
>   * FFprobe
>   
> You can also specify "PP+EXE:ARGS" to give the arguments to the specified executable only when being used by the specified postprocessor.
> 
> Additionally, for ffmpeg/ffprobe, "_i"/"_o" can be appended to the prefix optionally followed by a number to pass the argument before the specified input/output file. Eg: --ppa "Merger+ffmpeg_i1:-v quiet". You can use this option multiple times to give different arguments to different postprocessors. (Alias: --ppa)
> 

I have not added/modified any text barring the addition of newlines and asterisks.  One possible change could be to make the list of postprocessors a table of information, specifying which executable each postprocessor is applied by (assuming that they are _all_ handled by the specified executables and that none are bespoke functions with their own specific parameters), and listed (and explicitly stated to be) in the order they are applied, including whether argument order matters when postprocessors are specified explicitly.  A sentence or two describing each postprocessor could also be very useful.

Finally, examples of usage would also not go amiss, as I found @Zirro's above made it immediately obvious where I was wrong.

To conclude, and to be clear, I am not disputing that what I had was wrong, that I had made incorrect assumptions, and that it was 'user error', etc.  What I am saying is that I did attempt to seek out the correct approach, as well as trying to make this Issue as helpful and as clear as possible, and that I feel the documentation for this could be improved to avoid such issues in the future.

**Comment by pukkandan** [comment]
Just to clarify, the intend of my last comment was not to blame you. Just explaining why it didn't occur to me to recommend `ExtractAudio:` to you

**Comment by pukkandan** [comment]
> First, separate paragraphs would be extremely useful, as right now it's just a wall of text, which makes it easy to miss key nuggets of information held within.

The "usage and options" section of the readme cannot be formatted any other way. And it is not practical to create a new section for each option. So dedicated sections are generally only given to options like output template, format selection etc which needs large amounts of documentation

> Second, it is not mentioned which postprocessors are apparently applied by which executables

Correct, that is a implementation detail. Documenting it would imply that it is stable - but this may change in future without notice.


> nor is it mentioned that specifying the postprocessing options for an executable applies to all the postprocessors which run that executable.

~~What else did you expect to happen?~~

Edit: You understood it as "run the specified postprocessor"?

> I had assumed they were being listed as supported as separate additions to the postprocessors that could be configured, not that they were also handling the other postprocessing operations as well, and that they could be configured en-mass by specifying just the executable, as I had done.

The docs say "Give these arguments to the postprocessors", not "Run these additional postprocessors". That would be `--use-postprocessor`

> It's also unclear if a postprocessor operation can involve more than one executable.

It can. But what exe each PP uses is again an implementation detail and users should not depend on it

> I had also previously assumed that the specified postprocessors were all unique operations specific to yt-dlp (ie not separate executables), and with no supported arguments for each specified in the documentation, I had no idea they used what arguments could be applied to them.

Gotcha. Any suggestions on how we can reword it to make this clear.

> Once this information is known to someone, it may seem obvious to them, however to the initiated, such as myself, it seemed somewhat opaque.

Yes, this is one of the biggest issues with writing documentation! These things look obvious to me and can't often understand how it can be confusing until someone points it out

> Third, the documentation does not mention the order the post processors are run in. Does the order the parameters are passed in matter? What about the order of the ones that run automatically, such as M4AFixup? Or is the order always fixed regardless of whether the user manually specifies yt-dlp postprocessor parameters (such as --extract-audio) or not, or what order those parameters are passed in?

The order is always fixed. A specific order is chosen in such a way that the options don't conflict each other. For example, if `--remove-chapters` (`ModifyChapters`) is run after `--add-metadata` (`Metadata`), the added metadata will be lost. So yt-dlp's order does the reverse.

But the actual order in which these are run may change as new PPs are implemented or as some of the PPs are enhanced. So user must not depend on a specific order

> With regards to my first point regarding paragraphs, as an example of how much more readable paragraphs can make this, I have demonstrated below:

As I mentioned above, the usage and options section cannot be re-formatted. But a new section can be added for Postprocessors that goes into detail about the various postprocessing options (not just --post-processor-args). Would you be willing to make a PR adding this?

**Comment by Metritutus** [comment]
> > First, separate paragraphs would be extremely useful, as right now it's just a wall of text, which makes it easy to miss key nuggets of information held within.
> 
> The "usage and options" section of the readme cannot be formatted any other way. And it is not practical to create a new section for each option. So dedicated sections are generally only given to options like output template, format selection etc which needs large amounts of documentation

In that case, as you mention at the end of your post, a new section in the documentation sounds like the best approach for this.  The existing options section could then be amended to a 'check the documentation for further details and examples', or something along those lines.

> > Second, it is not mentioned which postprocessors are apparently applied by which executables
> 
> Correct, that is a implementation detail. Documenting it would imply that it is stable - but this may change in future without notice.

Whilst that is an implementation detail as you describe, it is also publicly configurable functionality due to the exisistence of `--postprocessor-args`.  This is unfortunate because it means the behaviour of a configured postprocessor configured by the user could also change without notice.

I would therefore feel that this information should still be listed, but could potentially include a disclaimer that it may not be fully up-to-date due because there could be postprocessor changes without notice, although it hopefully wouldn't be too difficult to keep it up-to-date.

> > nor is it mentioned that specifying the postprocessing options for an executable applies to all the postprocessors which run that executable.
> 
> ~What else did you expect to happen?~
> 
> Edit: You understood it as "run the specified postprocessor"?

I did!  At the time I understood it to be 'run this postprocessor with these arguments'.  My understanding has since been corrected on this as a result of this Issue.

> > I had assumed they were being listed as supported as separate additions to the postprocessors that could be configured, not that they were also handling the other postprocessing operations as well, and that they could be configured en-mass by specifying just the executable, as I had done.
> 
> The docs say "Give these arguments to the postprocessors", not "Run these additional postprocessors". That would be `--use-postprocessor`
> 
> > It's also unclear if a postprocessor operation can involve more than one executable.
> 
> It can. But what exe each PP uses is again an implementation detail and users should not depend on it

Considering that it affects configuration functionality that is accessible to the user, I feel it should still be documented.  A message can be included indicating that users shouldn't depend on this behaviour.

> > I had also previously assumed that the specified postprocessors were all unique operations specific to yt-dlp (ie not separate executables), and with no supported arguments for each specified in the documentation, I had no idea they used what arguments could be applied to them.
> 
> Gotcha. Any suggestions on how we can reword it to make this clear.

I'll give it some thought in relation to the potential pull-request for adding/improving to the documentation here.

> The order is always fixed. A specific order is chosen in such a way that the options don't conflict each other. For example, if `--remove-chapters` (`ModifyChapters`) is run after `--add-metadata` (`Metadata`), the added metadata will be lost. So yt-dlp's order does the reverse.
> 
> But the actual order in which these are run may change as new PPs are implemented or as some of the PPs are enhanced. So user must not depend on a specific order

It would still be useful to document this, along with the appropriate warnings.

> As I mentioned above, the usage and options section cannot be re-formatted. But a new section can be added for Postprocessors that goes into detail about the various postprocessing options (not just --post-processor-args). Would you be willing to make a PR adding this?

I'd certainly be willing to do so, at least with regards to paragraph formatting and so on for the new section, however I don't currently have sufficient knowledge to describe all the details, although I'll give it a shot (but may undoubtedly need to ask clarification questions regarding functionality of specific post processors), but I can certainly attempt to lay down some of the groundwork.

**Comment by pukkandan** [comment]
> but could potentially include a disclaimer that it may not be fully up-to-date due because there could be postprocessor changes without notice, although it hopefully wouldn't be too difficult to keep it up-to-date.

This is a bad idea imho and frankly an unnecessary overhead. `--ppa` is an advanced option and not meant for noobs. If a user cannot determine how the PP is working from debug logs, they likely shouldn't be using this option. How would they even know what option they want to pass if they don't understand what the PP is doing? 

There are also complex cases like `EmbedThumbnail` which may use ffmpeg, ffprobe, atomicparsley or none of them (or multiple of them) depending on what file extension is given and which of these tools are available. It is impossible to properly document these behaviors without literally explaining every step the PP takes. At that point, we are just making a human readable copy of the code...


> I'd certainly be willing to do so, at least with regards to paragraph formatting and so on for the new section, however I don't currently have sufficient knowledge to describe all the details, although I'll give it a shot

imo, `--exec`, `--use-postprocessor` and `--ppa` are the postprocessing options that would benefit most from additional documentation and examples (There is already a separate section for `--parse-metadata`/`--replace-in-metadata`). For the rest, we can ask to refer the usage options

Explanation of the "when" argument in `--use-postprocessor` is especially important because this is referenced in `--exec`, `--print` (and likely more options in the future)

> but may undoubtedly need to ask clarification questions regarding functionality of specific post processors

ofc, I'll help with clarifications and corrections

**Comment by marcello-pietrobon** [comment]
Just confirming that this thread has been very useful, and that it would be very useful to add some sections with practical examples on how to use --postprocessor-args , --use-postprocessor, and similar.

Beside that, the yt-dlp documentation is excellent.
